src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,114 +741,219 @@ private void ReturnReadBufferIfEmpty()
-                        // doesn't read enough), and to minimize the chances that in the common case the FillBufferAsync
-                        // helper needs to yield and allocate a state machine.
-                        int readBytes = await adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount)).ConfigureAwait(false);
-                        if (readBytes == 0)
+                        if (_internalBufferCount != 0)
                         {
-                            return 0;
+                            // we got EOF in middle of TLS frame. Treat that as error.
+                            throw new IOException(SR.net_io_eof);
                         }
 
-                        _internalBufferCount += readBytes;
-                        if (_internalBufferCount < SecureChannel.ReadHeaderSize)
-                        {
-                            await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize).ConfigureAwait(false);
-                        }
+                        return 0;
                     }
-                    Debug.Assert(_internalBufferCount >= SecureChannel.ReadHeaderSize);
 
-                    // Parse the frame header to determine the payload size (which includes the header size).
-                    int payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
-                    if (payloadBytes < 0)
+                    _internalBufferCount += bytesRead;
+                    if (frameSize == int.MaxValue && _internalBufferCount > SecureChannel.ReadHeaderSize)
                     {
-                        throw new IOException(SR.net_frame_read_size);
+                        frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
                     }
 
-                    // Read in the rest of the payload if we don't have it.
-                    if (_internalBufferCount < payloadBytes)
+                    if (_internalBufferCount < frameSize)
                     {
-                        await FillBufferAsync(adapter, payloadBytes).ConfigureAwait(false);
+                        t = adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount));
+                        continue;
                     }
 
-                    // Set _decrytpedBytesOffset/Count to the current frame we have (including header)
-                    // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
-                    _decryptedBytesOffset = _internalOffset;
-                    _decryptedBytesCount = payloadBytes;
+                    break;
+                }
+
+                return frameSize;
+            }
+        }
+
+        private SecurityStatusPal DecryptData(int frameSize, out int decryptedCount, out int decryptedOffset)
+        {
+            // Set _decrytpedBytesOffset/Count to the current frame we have (including header)
```

Comment:
Seems like we should assert that _decryptedBytesCount is 0 here, right? Otherwise we could be overwriting unconsumed decrypted data.

Reply:
yes, the asset is good idea.

Reply:
```suggestion
            // Set _decryptedBytesOffset/Count to the current frame we have (including header)
```